-Waggressive-loop-optimizations warning in FastNoiseLite thirdparty code (invoking undefined behavior)

### Godot version

4.0.dev (ef42b686adde506163f0a3702f5dc6fe80385da0)

### System information

Mageia 9 Linux, x86_64

### Issue description

When compiling with GCC 12 and `tools=yes target=release_debug use_lto=yes warnings=extra werror=yes`, I guess these warnings (treated as errors):

```
In file included from modules/noise/fastnoise_lite.h:39,
                 from modules/noise/fastnoise_lite.cpp:31:
./thirdparty/noise/FastNoiseLite.h: In member function 'fastnoiselite::FastNoiseLite::SingleCellular<float>(int, float, float, float) constfloat':
./thirdparty/noise/FastNoiseLite.h:1726:33: error: iteration 2 invokes undefined behavior [-Werror=aggressive-loop-optimizations]
 1726 |                         zPrimed += PrimeZ;
      |                         ~~~~~~~~^~~~~~~~~
./thirdparty/noise/FastNoiseLite.h:1709:46: note: within this loop
 1709 |                     for (int zi = zr - 1; zi <= zr + 1; zi++)
      |                                           ~~~^~~~~~~~~
./thirdparty/noise/FastNoiseLite.h:1693:33: error: iteration 2 invokes undefined behavior [-Werror=aggressive-loop-optimizations]
 1693 |                         zPrimed += PrimeZ;
      |                         ~~~~~~~~^~~~~~~~~
./thirdparty/noise/FastNoiseLite.h:1676:46: note: within this loop
 1676 |                     for (int zi = zr - 1; zi <= zr + 1; zi++)
      |                                           ~~~^~~~~~~~~
./thirdparty/noise/FastNoiseLite.h:1660:33: error: iteration 2 invokes undefined behavior [-Werror=aggressive-loop-optimizations]
 1660 |                         zPrimed += PrimeZ;
      |                         ~~~~~~~~^~~~~~~~~
./thirdparty/noise/FastNoiseLite.h:1643:46: note: within this loop
 1643 |                     for (int zi = zr - 1; zi <= zr + 1; zi++)
      |                                           ~~~^~~~~~~~~
```

It's thirdparty code so we'd have to patch it to solve it, and ideally send the patch upstream.

### Steps to reproduce

- `scons p=linuxbsd tools=yes target=release_debug use_lto=yes warnings=extra werror=yes` with GCC 12 (`gcc (Mageia 12.0.1-0.20220424.1.mga9) 12.0.1 20220424 (experimental)`)

### Minimal reproduction project

_No response_

--- a/thirdparty/noise/FastNoiseLite.h
+++ b/thirdparty/noise/FastNoiseLite.h
@@ -1611,6 +1611,12 @@ private:
         }
     }
 
+// GCC raises warnings when integer overflows occur, which are needed for hashing here.
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Waggressive-loop-optimizations"
+#endif
+
     template <typename FNfloat>
     float SingleCellular(int seed, FNfloat x, FNfloat y, FNfloat z) const
     {
@@ -1765,6 +1771,9 @@ private:
         }
     }
 
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
 
     // Perlin Noise
 

--- a/thirdparty/noise/patches/FastNoiseLite.patch
+++ b/thirdparty/noise/patches/FastNoiseLite.patch
@@ -430,3 +430,26 @@
      {
          x *= frequency;
          y *= frequency;
+@@ -1611,6 +1611,12 @@ private:
+         }
+     }
+ 
++// GCC raises warnings when integer overflows occur, which are needed for hashing here.
++#if defined(__GNUC__) && !defined(__clang__)
++#pragma GCC diagnostic push
++#pragma GCC diagnostic ignored "-Waggressive-loop-optimizations"
++#endif
++
+     template <typename FNfloat>
+     float SingleCellular(int seed, FNfloat x, FNfloat y, FNfloat z) const
+     {
+@@ -1765,6 +1771,9 @@ private:
+         }
+     }
+ 
++#if defined(__GNUC__) && !defined(__clang__)
++#pragma GCC diagnostic pop
++#endif
+ 
+     // Perlin Noise
+ 

